### ISSUE TITLE
High CPU until first event is written over IPC socket

### ISSUE THREAD

**Issue by sodface** [project-repo/cagebreak#46]
While testing scripts with the IPC socket I noticed that one core on my 4 core desktop goes to 100% and stays there until the first event occurs and is read by the script, at which point the CPU usage drops back down and the script and event processing seems to progress as expected from then on.

This also occurs when just connecting directly with nc (I tested with both busybox nc and openbsd-netcat), single core CPU goes to 100% until the first event occurs.

The process at 100% is `cagebreak -e` so it seems like a cagebreak issue at first glance but maybe it's something on the client side?

I'm a novice at troubleshooting this type of thing, but strace is logging a lot of `epoll_pwait` calls during high CPU and then those taper off dramatically after the first event and CPU usage goes down.

```
~ $ doas -c -f strace -p 6983
strace: Process 6983 attached with 11 threads
^Cstrace: Process 6983 detached
strace: Process 6984 detached
strace: Process 6985 detached
strace: Process 6986 detached
strace: Process 6987 detached
strace: Process 6988 detached
strace: Process 6989 detached
strace: Process 6990 detached
strace: Process 6991 detached
strace: Process 6992 detached
strace: Process 6993 detached
% time     seconds  usecs/call     calls    errors syscall
------ ----------- ----------- --------- --------- ----------------
 94.62    2.724267           1   2096153           epoll_pwait
  4.73    0.136285          20      6719        28 ioctl
  0.51    0.014639           5      2913       397 read
  0.09    0.002717           6       428           sendmsg
  0.01    0.000409           3       114           write
  0.01    0.000290           9        32           recvmsg
  0.01    0.000183          10        17           munmap
  0.01    0.000153           8        18           mmap
  0.01    0.000145           6        24           timerfd_settime
  0.00    0.000031           3         8           fcntl
  0.00    0.000014          14         1           accept
  0.00    0.000012           6         2           epoll_ctl
------ ----------- ----------- --------- --------- ----------------
100.00    2.879145           1   2106429       425 total
```

**Comment by project-repo** [comment]
Hi sodface

this certainly sounds worrying. However, we just tried to reproduce this
bug on two independent machines and were unable to do so.

Could you please state any available details on your setup to enable
reproducibility on our side (flags used for compiling cagebreak, OS,
exact command which was run, environment (e.g. TTY or nested within a
wayland session), etc.)?

cheers
project-repo

**Comment by sodface** [comment]
Hi project-repo, I just did a fresh install on the same hardware. Some details:

- Alpine Linux
- Cagebreak 2.1.1 ([See APKBUILD here](https://git.alpinelinux.org/aports/tree/community/cagebreak/APKBUILD))
- As for compiler flags... I'm not 100% sure, I'm using the package built from Alpine's builder, [the CI build log can viewed here](https://gitlab.alpinelinux.org/sodface/aports/-/jobs/1014117) if that helps at all.

I'm starting cagebreak from TTY. Contents of .profile:

```
#!/bin/sh

mkdir -m 0700 -p /tmp/user/$(id -u)
export GDK_BACKEND=wayland
export HISTSIZE=131172 # 128K lines
export TERMINAL=sakura
export XDG_RUNTIME_DIR=/tmp/user/$(id -u)
export MOZ_ENABLE_WAYLAND=1
export EGL_PLATFORM=wayland

if [ -z "$DISPLAY" ] && [ "$(tty)" = "/dev/tty1" ]; then
	date +%s%3N > starttime.txt
	exec dbus-run-session -- cagebreak -e
fi
```
The screen recording shows my second monitor with 1 split, htop above and then connecting to IPC with busybox nc below. `cagebreak -e` goes to 100% until I move my mouse from the bottom tile to the top causing an event, at which time CPU usage drops off.

https://user-images.githubusercontent.com/46695384/234120547-66c5534c-a4df-4028-b37a-a9d7546bb94b.mp4

**Comment by nekopsykose** [comment]
i can't get any readable callgraph, but more or less it starts in wlroots

**Comment by project-repo** [comment]
Hi everyone,
Thanks a lot for the help with debugging and additional system
information. We are now able to reproduce this and pushed a tentative
fix for it yesterday (see `development` branch). Can you confirm that this fixes the issue?
Cheers,
project-repo

**Comment by nekopsykose** [comment]
fixes it for me in my testing

**Comment by sodface** [comment]
I'm not as fast as psykose but I'll test asap. Thanks!

**Comment by sodface** [comment]
Fixes it for me too. Thanks again!

**Comment by project-repo** [comment]
Fixed in 2.2.0
